### PR TITLE
8314327: Issues with JShell when using "local" execution engine

### PR DESCRIPTION
**Pull request by archiecobbs**
There are a few bugs (or "opportunities for improvement") in JShell's support for "local" execution control.

**Background**

JShell utilizes two distinct "class lookup" functions. First, it looks up clases when compiling snippets. This happens indirectly via the Java compiler using its `JavaFileSystem` interface. Secondly, JShell executes code by supplying an execution engine with class files and asking it to invoke methods. The execution engine therefore performs a class lookup function when it links and loads these classes. The compilation happens on the local machine while the execution happens either locally or remotely.

For everything to run smoothly, both of these lookup functions must be working for whatever class(es) you want to name in JShell input.

Local execution is implemented by `LocalExecutionControl` and provided by the `LocalExecutionControlProvider` under the name `"local"`. In turn, `LocalExecutionControl` uses a `LoaderDelegate` to manage class loading. The default `DefaultLoaderDelegate` uses a custom subclass of `URLClassLoader` (called `RemoteClassLoader`). This class loader has two functions: (a) keep track of snippet classes generated by JShell, and (b) find classes for the execution engine (normal class loader function).

**Issue 1**

The `RemoteClassLoader` is hard-wired with the system class loader as its parent loader. This means that in non-trivial class loading scenarios, such as when running in web servlet container, the application's classes are not visible to the local execution engine.  As a result, it's impossible to resolve any of these classes.

**Issue 2**

JShell has a `--class-path` parameter that is supposed to handle properly configuring *both* lookup functions, i.e., compilation and execution. Internally, this is done by (a) configuring the compiler's class path, and (b) passing this flag to the execution engine as a "remove VM option".

However, the "local" execution engine ignores the remote VM options.

Here's a simple demonstration:

```
    $ ls classes/test/Foo.class
    classes/test/Foo.class
    $ jshell --execution=local --class-path classes
    |  Welcome to JShell -- Version 20.0.1
    |  For an introduction type: /help intro

    jshell> Class.forName("test.Foo");
    |  Exception java.lang.ClassNotFoundException: test.Foo
    |        at URLClassLoader.findClass (URLClassLoader.java:445)
    |        at DefaultLoaderDelegate$RemoteClassLoader.findClass (DefaultLoaderDelegate.java:154)
    |        at ClassLoader.loadClass (ClassLoader.java:588)
    |        at ClassLoader.loadClass (ClassLoader.java:521)
    |        at Class.forName0 (Native Method)
    |        at Class.forName (Class.java:391)
    |        at Class.forName (Class.java:382)
    |        at (#1:1)

    jshell> Thread.currentThread().getContextClassLoader();
    $2 ==> jdk.jshell.execution.DefaultLoaderDelegate$RemoteClassLoader@45283ce2

    jshell> Arrays.asList(((URLClassLoader)$2).getURLs())
    $3 ==> []

    jshell> new File("classes/test/Foo.class").exists()
    $4 ==> true
```

**Issue 3**

JShell defines a `LoaderDelegate` interface, and provides a perfectly reasonable default implementation in `DefaultLoaderDelegate`, but makes it impossible for anyone wishing to customize `LocalExecutionControl` to leverage any of the existing `DefaultLoaderDelegate` implementation, because the class is not public. Also, this class is non-trivial and so duplicating its tricky class loading function is tedious and error-prone.

There's no reason not to make `DefaultLoaderDelegate` a public class, because (a) it adds no API surface beyond what is already defined by the already-public `LoaderDelegate` interface, and (b) its current behavior is already being relied upon (it's always used), so by making it public no additional dependencies or expectations would be created beyond those that already exist.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Change requires CSR request [JDK-8314328](https://bugs.openjdk.org/browse/JDK-8314328) to be approved
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8314327](https://bugs.openjdk.org/browse/JDK-8314327): Issues with JShell when using "local" execution engine (**Bug** - P4)
 * [JDK-8314328](https://bugs.openjdk.org/browse/JDK-8314328): Support configuring LocalExecutionControl's parent ClassLoader (**CSR**)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15311/head:pull/15311` \
`$ git checkout pull/15311`

Update a local copy of the PR: \
`$ git checkout pull/15311` \
`$ git pull https://git.openjdk.org/jdk.git pull/15311/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15311`

View PR using the GUI difftool: \
`$ git pr show -t 15311`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15311.diff">https://git.openjdk.org/jdk/pull/15311.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15311#issuecomment-1680786858)

### CODE REVIEW

**Comment by archiecobbs** [comment]
@lahodaj would you mind reviewing the CSR? [JDK-8314328](https://bugs.openjdk.org/browse/JDK-8314328)

Thanks.

**Comment by archiecobbs** [comment]
> Having three different APIs that only avoid implementing custom `LoaderDelegate` seems a bit excessive to me, unless there's a strong reason for them. I would suggest to settle on one way, and delete the others. Offhand, my preference order would be:
> 
>     * new `LocalExecutionControl` constructor (it is the smallest API surface, so as long as it solves the problem, it seems like the obvious solution)
> 
>     * making `DefaultLoaderDelegate` public (the behavioral compatibility we might need to keep could be non-trivial)
> 
>     * using context CL (I don't quite see the benefit of this, if we can add satisfactory real API, and the API surface seems pretty big, so we are not really saving anything)

I agree it makes sense to keep the API changes minimal. The problem is that the `--class-path` fix entails adding the path components to whatever `LoaderDelegate` you are using via `addToClasspath()`, so the two problems we're trying to solve intersect.

In other words, consider your first option:
* It requires you to write your own `ExecutionControlProvider`
* The fix for the `--class-path` bug is part of `LocalContextExecutionControlProvider`
* There currently no way to subclass `LocalContextExecutionControlProvider` and also inherit that fix

So you're stuck with choosing between being able to supply a parent `ClassLoader` or having the `--class-path` bug fixed, but not both (or duplicating the bug fix in your own implementation).

Here's another way it could work:

* Start with your first option, i.e., add the new `LocalExecutionControl` constructor
* Making these two new methods in the patch for `LocalExecutionControlProvider` protected and therefore overriddable by subclasses:
  * `createLoaderDelegate()`
  * `createExecutionControl()`

Of course that also adds API surface.

An alternative could be encapsulating the bug fix in its own static method in `LocalExecutionControlProvider`:
```java
public static void applyRemoteVMOptions(ExecutionEnv env, LoaderDelegate delegate) { ... }
```
or even as a default method in `LoaderDelegate`:
```java
default void initialize(ExecutionEnv env, Map<String, String> parameters) { ... }
```

File: test/langtools/jdk/jshell/LocalExecutionTestSupport.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import org.testng.annotations.BeforeTest;
+
+/*
+ * This class installs a class in a temporary diretory so we can test
+ * finding classes that are not visible to the system class loader.
+ */
+public class LocalExecutionTestSupport extends ReplToolTesting {
+
+    /*
+     * This is a classfile corresponding to this source:
```
[the remark this answers is not in the document]
Comment (archiecobbs) [reply test/langtools/jdk/jshell/LocalExecutionTestSupport.java]: Thanks - fixed in 50bb7612c16.

**Comment by archiecobbs** [comment]
@lahodaj any opinions on the best way to proceed? Thanks for taking the time to review.

**Comment by lahodaj** [comment]
Seems adding `createExecutionControl()` to `LocalExecutionControlProvider` should be enough? Something along these lines possibly?
```
diff --git a/src/jdk.jshell/share/classes/jdk/jshell/execution/LocalExecutionControlProvider.java b/src/jdk.jshell/share/classes/jdk/jshell/execution/LocalExecutionControlProvider.java
index cc3879c5c6d..370fbb6bdfd 100644
--- a/src/jdk.jshell/share/classes/jdk/jshell/execution/LocalExecutionControlProvider.java
+++ b/src/jdk.jshell/share/classes/jdk/jshell/execution/LocalExecutionControlProvider.java
@@ -25,6 +25,7 @@
 
 package jdk.jshell.execution;
 
+import java.util.List;
 import java.util.Map;
 import jdk.jshell.spi.ExecutionControl;
 import jdk.jshell.spi.ExecutionControlProvider;
@@ -78,7 +79,31 @@ public Map<String,String> defaultParameters() {
      */
     @Override
     public ExecutionControl generate(ExecutionEnv env, Map<String, String> parameters) {
-        return new LocalExecutionControl();
+        ExecutionControl result = createExecutionControl(env, parameters);
+
+        // Apply any configured class path
+        List<String> remoteOptions = env.extraRemoteVMOptions();
+        int classPathIndex = remoteOptions.indexOf("--class-path") + 1;
+        if (classPathIndex > 0 && classPathIndex < remoteOptions.size()) {
+            try {
+                result.addToClasspath(remoteOptions.get(classPathIndex));
+            } catch (ExecutionControl.ExecutionControlException e) {
+                throw new RuntimeException("error configuring class path", e);
+            }
+        }
+
+        return result;
     }
 
+    /**
+     * Create the {@link ExecutionControl} using the specified delegate.
+     * Called from {@link #generate(ExecutionEnv, Map) }.
+     *
+     * @param env the {@code ExecutionEnv} for which the ExecutionControl should be created
+     * @param parameters the parameters that were passed to {@code generate}
+     * @return the newly created {@code ExecutionControl}
+     */
+    public ExecutionControl createExecutionControl(ExecutionEnv env, Map<String, String> parameters) {
+        return new LocalExecutionControl(new DefaultLoaderDelegate());
+    }
 }
```

**Comment by archiecobbs** [comment]
> Seems adding createExecutionControl() to LocalExecutionControlProvider should be enough? Something along these lines possibly?

Ah yes, that would work... because `DirectExecutionControl.addToClasspath()` invokes `LoaderDelegate.addToClasspath()`.

It's better place to apply the class path at that higher level anyway.

Thanks! I've applied this in cab048a87a3 ... let me know what you think.

**Review by lahodaj — APPROVED** [review]
Looks reasonable to me.

**Comment by johnhart-s1** [comment]
@archiecobbs Thanks for this work; any chance this will get sponsored & merged?

**Comment by archiecobbs** [comment]
> @archiecobbs Thanks for this work; any chance this will get sponsored & merged?

I hope so...   @lahodaj any remaining issues?

**Comment by lahodaj** [comment]
Thanks for sponsoring the PR, Vicente. I apologize, I forgot about it.

**Pull request closed without merging**